Generating strongly typed resource files not currently supported on .NET Core MSBuild

_From @iSazonov on July 6, 2017 13:41_

What is a right repo to ask about ResGen in `dotnet build`?

Currently I try to generate strongly typed resources and get error:
```C:\Users\user\AppData\Local\Microsoft\dotnet\sdk\2.0.0-preview2-006502\Microsoft.Common.CurrentVersion.targets(2921,5): error : Generating strongly typed resource files not currently supported on .NET Core MSBuild [C:\Users\user\Documents\GitHub\iSazonov\PowerShell\src\Microsoft.PowerShell.CoreCLR.Eventing\Microsoft.PowerShell.CoreCLR.Eventing.csproj]```

From related Issue https://github.com/dotnet/cli/issues/819:
> dotnet-resgen  got replaced by the  Microsoft.Net.Sdk  implementation, so closing this.

Is strongly typed resources supported?

_Copied from original issue: dotnet/cli#7084_

@danmosemsft, I tried building msbuild master on Windows 10, `msbuild\build.cmd -hostType Core` and getting errors like:

> C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at LibGit2Sharp.Core.Proxy.git_revwalk_next(RevWalkerHandle walker) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\TaskUsageLogger\TaskUsageLogger.csproj]

Tried in three modes: Developer Command Prompt VS2017, normal cmd, normal PowerShell.

I just cloned the repo, tip of master branch. Call stack:

```
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly. [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018: LibGit2Sharp.NotFoundException: object not found - no match for id (7c1f7a8cd8f2b844046892036912db28a7ec4c37) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at LibGit2Sharp.Core.Proxy.git_revwalk_next(RevWalkerHandle walker) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at LibGit2Sharp.CommitLog.CommitEnumerator.MoveNext() [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at Nerdbank.GitVersioning.VersionOracle..ctor(String projectDirectory, Repository repo, ICloudBuild cloudBuild, Nullable`1 overrideBuildNumberOffset, String projectPathRelativeToGitRepoRoot) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at Nerdbank.GitVersioning.VersionOracle.Create(String projectDirectory, String gitRepoDirectory, ICloudBuild cloudBuild, Nullable`1 overrideBuildNumberOffset, String projectPathRelativeToGitRepoRoot) [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
C:\Users\kasper\.nuget\packages\nerdbank.gitversioning\2.1.23\build\Nerdbank.GitVersioning.targets(63,5): error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner() [C:\Users\kasper\Source\Repos\msbuild\src\Samples\XmlFileLogger\XmlFileLogger.csproj]
```

```
git log -1
commit e8b480c373eadd9daa975c6e96e4dbbabcd9a4fe
```

--- a/src/Tasks/ResourceHandling/MSBuildResXReader.cs
+++ b/src/Tasks/ResourceHandling/MSBuildResXReader.cs
@@ -82,6 +82,7 @@ namespace Microsoft.Build.Tasks.ResourceHandling
         private const string ResMimeType = "text/microsoft-resx";
 
         private const string StringTypeName = "System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089";
+        private const string MemoryStreamTypeNameDesktopFramework = "System.IO.MemoryStream, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089";
 
         private static string GetFullTypeNameFromAlias(string aliasedTypeName, Dictionary<string, string> aliases)
         {
@@ -157,6 +158,7 @@ namespace Microsoft.Build.Tasks.ResourceHandling
             {
                 if (IsByteArray(typename))
                 {
+                    // Handle byte[]'s, which are stored as base-64 encoded strings.
                     byte[] byteArray = Convert.FromBase64String(value);
 
                     resources.Add(new LiveObjectResource(name, byteArray));
@@ -232,14 +234,19 @@ namespace Microsoft.Build.Tasks.ResourceHandling
             }
             else if (IsByteArray(fileRefType))
             {
-                using (FileStream s = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read))
-                {
-                    byte[] byteArray = new byte[s.Length];
-                    s.Read(byteArray, 0, (int)s.Length);
+                byte[] byteArray = File.ReadAllBytes(fileName);
 
-                    resources.Add(new LiveObjectResource(name, byteArray));
-                    return;
-                }
+                resources.Add(new LiveObjectResource(name, byteArray));
+                return;
+            }
+            else if (IsMemoryStream(fileRefType))
+            {
+                // See special-case handling in ResXFileRef
+                // https://github.com/dotnet/winforms/blob/689cd9c69e632997bc85bf421af221d79b12ddd4/src/System.Windows.Forms/src/System/Resources/ResXFileRef.cs#L293-L297
+                byte[] byteArray = File.ReadAllBytes(fileName);
+
+                resources.Add(new LiveObjectResource(name, new MemoryStream(byteArray)));
+                return;
             }
 
             resources.Add(new FileStreamResource(name, fileRefType, fileName, resxFilename));
@@ -249,14 +256,19 @@ namespace Microsoft.Build.Tasks.ResourceHandling
         /// Does this assembly-qualified type name represent an array of bytes?
         /// </summary>
         /// <remarks>
+        /// We can't hard-code byte[] type name due to version number
+        /// updates & potential whitespace issues with ResX files.
+        ///
         /// Comment and logic from https://github.com/dotnet/winforms/blob/16b192389b377c647ab3d280130781ab1a9d3385/src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs#L411-L416
         /// </remarks>
-        // Handle byte[]'s, which are stored as base-64 encoded strings.
-        // We can't hard-code byte[] type name due to version number
-        // updates & potential whitespace issues with ResX files.
         private static bool IsByteArray(string fileRefType)
         {
             return fileRefType.IndexOf("System.Byte[]") != -1 && fileRefType.IndexOf("mscorlib") != -1;
+        }
+
+        private static bool IsMemoryStream(string fileRefType)
+        {
+            return fileRefType.Equals(MemoryStreamTypeNameDesktopFramework, StringComparison.Ordinal);
         }
 
         /// <summary>

--- a/src/Tasks/ResourceHandling/MSBuildResXReader.cs
+++ b/src/Tasks/ResourceHandling/MSBuildResXReader.cs
@@ -230,6 +230,17 @@ namespace Microsoft.Build.Tasks.ResourceHandling
                     return;
                 }
             }
+            else if (IsByteArray(fileRefType))
+            {
+                using (FileStream s = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read))
+                {
+                    byte[] byteArray = new byte[s.Length];
+                    s.Read(byteArray, 0, (int)s.Length);
+
+                    resources.Add(new LiveObjectResource(name, byteArray));
+                    return;
+                }
+            }
 
             resources.Add(new FileStreamResource(name, fileRefType, fileName, resxFilename));
         }

--- a/src/Tasks/ResourceHandling/MSBuildResXReader.cs
+++ b/src/Tasks/ResourceHandling/MSBuildResXReader.cs
@@ -155,14 +155,10 @@ namespace Microsoft.Build.Tasks.ResourceHandling
 
             if (mimetype == null)
             {
-                if (typename.IndexOf("System.Byte[]") != -1 && typename.IndexOf("mscorlib") != -1)
+                if (IsByteArray(typename))
                 {
                     byte[] byteArray = Convert.FromBase64String(value);
 
-                    // Comment and logic from https://github.com/dotnet/winforms/blob/16b192389b377c647ab3d280130781ab1a9d3385/src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs#L411-L416
-                    // Handle byte[]'s, which are stored as base-64 encoded strings.
-                    // We can't hard-code byte[] type name due to version number
-                    // updates & potential whitespace issues with ResX files.
                     resources.Add(new LiveObjectResource(name, byteArray));
                     return;
                 }
@@ -236,6 +232,20 @@ namespace Microsoft.Build.Tasks.ResourceHandling
             }
 
             resources.Add(new FileStreamResource(name, fileRefType, fileName, resxFilename));
+        }
+
+        /// <summary>
+        /// Does this assembly-qualified type name represent an array of bytes?
+        /// </summary>
+        /// <remarks>
+        /// Comment and logic from https://github.com/dotnet/winforms/blob/16b192389b377c647ab3d280130781ab1a9d3385/src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs#L411-L416
+        /// </remarks>
+        // Handle byte[]'s, which are stored as base-64 encoded strings.
+        // We can't hard-code byte[] type name due to version number
+        // updates & potential whitespace issues with ResX files.
+        private static bool IsByteArray(string fileRefType)
+        {
+            return fileRefType.IndexOf("System.Byte[]") != -1 && fileRefType.IndexOf("mscorlib") != -1;
         }
 
         /// <summary>

